text/3519-arbitrary-self-types-v2.md

```diff
@@ -83,20 +83,55 @@ See also [this blog post](https://medium.com/@adetaylor/the-case-for-stabilizing
 
 ## Motivation for the v2 changes
 
-Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`.
+Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`. Example:
 
-However, if it's OK to create a reference `&T`, you _probably don't need this feature_. You can probably simply use `&self` as your receiver type.
+```rust
+#[feature(arbitrary_self_types)]
+
+struct SmartPtr<T>(*const T);
+
+impl<T: ?Sized> Deref for SmartPtr<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        // never called, but smart pointers need to implement this method
+        // sometimes it's just not safe to create a reference to self.0
+    }
+}
+
+struct ConcreteType;
+
+impl ConcreteType {
+    fn some_method(self: SmartPtr<ConcreteType>) {
+
+    }
+}
+
+fn main() {
+    let concrete: SmartPtr<ConcreteType> = ...;
+    concrete.some_method();
+}
+```
+
+However, if it's OK to create a reference `&T`, you _probably_ don't need this feature. You can simply use `&self` as your receiver type:
+
+```rust
+impl ConcreteType {
+    fn some_method(&self) {
+
+    }
+}
+```
 
-This feature is fundamentally aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.
+This feature is mostly aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.
```

Comment:
I wouldn't agree with the "mostly" here. The biggest motivation for this feature was to call methods on `Pin<&mut Self>` -- methods like `Future::poll`; it is no coincidence that this was stabilized together with the entire `async`/`Future` push. `Rc<Self>` was also important.

Supporting unsafe raw pointers is a nice addition, but arbitrary self receivers is very useful even in its current form.

Reply:
OK. Thanks. I'll think about how to reword this.

Reply:
OK, I adjusted this in e269c006. I'm going to resolve this conversation so I can keep track of what still needs to be addressed, but please do reopen it if you don't think it accurately reflects the situation.